File: 20.py
"""
https://leetcode.com/problems/valid-parentheses/
"""

pattern = {
    "(": ")",
    "[": "]",
    "{": "}",
}


class Solution:
    def isValid(self, s: str) -> bool:
        stack = list()

        for c in s:
            if c in pattern.keys():
                stack.append(pattern[c])
            elif not stack or stack.pop() != c:
                return False

        if stack:
            return False
        return True


s = Solution()

# print(s.isValid("()"))
# print(s.isValid("()[]{}"))
# print(s.isValid("(]"))
print(s.isValid("]"))

File: leetcode/20.go
package main

import "fmt"

var pattern = map[rune]rune{
	'(': ')',
	'[': ']',
	'{': '}',
}

func isValid(s string) bool {
	stack := make([]rune, 0)
	for _, c := range s {
		if val, ok := pattern[c]; ok {
			stack = append(stack, val)
		} else if len(stack) == 0 || stack[len(stack)-1] != c {
			return false
		} else {
			stack = stack[:len(stack)-1] // Correctly pop the element from the stack
		}

	}
	return len(stack) == 0
}

func main() {
	fmt.Println(isValid("()"))     // true
	fmt.Println(isValid("()[]{}")) // true
	fmt.Println(isValid("(]"))     // false
	fmt.Println(isValid("]"))      // false
}
